Tools/clinic/clinic.py

```diff
@@ -535,56 +535,65 @@ def normalize_snippet(s, *, indent=0):
     return s
 
 
-def declare_parser(*, hasformat=False):
+def declare_parser(f, *, hasformat=False):
     """
     Generates the code template for a static local PyArg_Parser variable,
     with an initializer.  For core code (incl. builtin modules) the
     kwtuple field is also statically initialized.  Otherwise
     it is initialized at runtime.
     """
+    def keywords(f):
+        params = f.parameters.values()
+        kwds = [
+            p for p in params
+            if not p.is_positional_only() and not p.is_vararg()
+        ]
+        return kwds
+
     if hasformat:
         fname = ''
         format_ = '.format = "{format_units}:{name}",'
     else:
         fname = '.fname = "{name}",'
         format_ = ''
-    declarations = """
-        #define NUM_KEYWORDS {num_keywords}
-        #if NUM_KEYWORDS == 0
-
-        #  if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)
-        #    define KWTUPLE (PyObject *)&_Py_SINGLETON(tuple_empty)
-        #  else
-        #    define KWTUPLE NULL
-        #  endif
-
-        #else  // NUM_KEYWORDS != 0
-        #  if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)
-
-        static struct {{
-            PyGC_Head _this_is_not_used;
-            PyObject_VAR_HEAD
-            PyObject *ob_item[NUM_KEYWORDS];
-        }} _kwtuple = {{
-            .ob_base = PyVarObject_HEAD_INIT(&PyTuple_Type, NUM_KEYWORDS)
-            .ob_item = {{ {keywords_py} }},
-        }};
-        #  define KWTUPLE (&_kwtuple.ob_base.ob_base)
-
-        #  else  // !Py_BUILD_CORE
-        #    define KWTUPLE NULL
-        #  endif  // !Py_BUILD_CORE
-        #endif  // NUM_KEYWORDS != 0
-        #undef NUM_KEYWORDS
-
-        static const char * const _keywords[] = {{{keywords_c} NULL}};
-        static _PyArg_Parser _parser = {{
-            .keywords = _keywords,
-            %s
-            .kwtuple = KWTUPLE,
-        }};
-        #undef KWTUPLE
-        """ % (format_ or fname)
+
+    num_keywords = len(keywords(f))
```

Comment:
FWIW, you could do away with `keywords()`:

```suggestion
    num_keywords = len([p for p in f.parameters.values()
                        if not p.is_positional_only() and not p.is_vararg()])
```

Reply:
Yeah, the initial draft I had used something like that. I added the nested function for readability. I have no strong preference here; I'm fine with it as it is, and I'm also fine with inlining it.